src/index.ts

```diff
@@ -369,39 +399,83 @@ class ThreadPool {
+    this._onWorkerAvailable(workerInfo);
+  }
+
   _onWorkerAvailable (workerInfo : WorkerInfo) : void {
-    while ((this.taskQueue.size > 0 || this.skipQueue.length > 0) &&
-      workerInfo.currentUsage() < this.options.concurrentTasksPerWorker) {
+    let workers: PiscinaWorker[] | null = null;
+    while ((this.taskQueue.size > 0 || this.skipQueue.length > 0)) {
       // The skipQueue will have tasks that we previously shifted off
       // the task queue but had to skip over... we have to make sure
       // we drain that before we drain the taskQueue.
       const taskInfo = this.skipQueue.shift() ||
                        this.taskQueue.shift() as TaskInfo;
-      // If the task has an abortSignal and the worker has any other
-      // tasks, we cannot distribute the task to it. Skip for now.
-      if (taskInfo.abortSignal && workerInfo.taskInfos.size > 0) {
-        this.skipQueue.push(taskInfo);
+
+      if (workers == null) {
+        workers = [...this.workers].map(workerInfo => workerInfo.interface);
+      }
+
+      const distributed = this._distributeTask(taskInfo, workers);
+
+      if (distributed) {
+        // If task was distributed, we should continue to distribute more tasks
+        continue;
+      } else {
+        // If balancer states that pool is busy, we should stop trying to distribute tasks
         break;
       }
-      const now = performance.now();
-      this.waitTime?.record(toHistogramIntegerNano(now - taskInfo.created));
-      taskInfo.started = now;
-      workerInfo.postTask(taskInfo);
-      this._maybeDrain();
-      return;
     }
 
-    if (workerInfo.taskInfos.size === 0 &&
+    // If more workers than minThreads, we can remove idle workers
+    if (workerInfo.currentUsage() === 0 &&
         this.workers.size > this.options.minThreads) {
       workerInfo.idleTimeout = setTimeout(() => {
-        assert.strictEqual(workerInfo.taskInfos.size, 0);
+        assert.strictEqual(workerInfo.currentUsage(), 0);
         if (this.workers.size > this.options.minThreads) {
           this._removeWorker(workerInfo);
         }
       }, this.options.idleTimeout).unref();
     }
   }
 
+  _distributeTask (task: TaskInfo, workers: PiscinaWorker[]): boolean {
+    // We need to verify if the task is aborted already or not
+    // otherwise we might be distributing aborted tasks to workers
+    if (task.aborted) return false;
+
+    const candidate = this.balancer(task.interface, workers);
+
+    // Seeking for a real worker instead of customized one
+    if (candidate != null && candidate[kWorkerData] != null) {
+      const now = performance.now();
+      this.waitTime?.record(toHistogramIntegerNano(now - task.created));
+      task.started = now;
+      candidate[kWorkerData].postTask(task);
+      this._maybeDrain();
+      // If candidate, let's try to distribute more tasks
+      return true;
+    }
+
+    // We spawn if possible
+    // TODO: scheduler will intercept this.
```

Comment:
What do you mean here? I think worker creation creation should be done before calling the balancer: the newly created worker is flagged ready and usable.

Reply:
Good call

Overall, the current state is to just spawn new workers based on balancer feedback.

How it right now works is:
- Based on `minThreads` and `maxThreads` Piscina will spawn the necessary workers
- On every task distributed or enqueued, we will check `maxThreads` to verify if we can create more workers

In the future, how I envision this is:
- Based on `minThreads` and `maxThreads` Piscina will spawn the necessary workers
- When a task is about to be distributed
  - Ask the scheduler to verify if we should spawn more workers (and spawn if so)
  - Call the balancer to distribute the workload
  - If not possible to distribute, we enqueue